### PR TITLE
[cxx-interop] Lazily import members of clang namespaces and records via requests.

### PR DESCRIPTION
**Pull request by zoecarver**
This patch does three things: 

1. It makes namespaces lazy and transparent. What I mean by this is that all members will be loaded lazily. Also, namespaces will always appear to be empty (i.e., have no members). Whenever a member is requested, the specific member must be requested. This prevents us from accidently doing things we shouldn't. Namespaces themselves aren't physical things and shouldn't appear to exist or hold anything. 

2. It makes records lazy. They will now lazily load their members, but will also hold on to the loaded members unlike namespaces. As time goes on I want to move more stuff out of the record visitor but I'll do that in follow up patches.

3. We now print things in the __ObjC module along with whatever module was requested.

As a meta-note: while there are about the same number of lines added as removed, this code is now a lot simpler. This turned many loops with ad-hoc conditions into one loop in a request. It also composes a bit better: the logic for printing namespaces isn't in the namespace visitor it's in the ASTPrinter. The logic for bailing on fields is in the field visitor not the record visitor. Etc. 

As more people start working on C++ interop, I'm hoping to simplify the code more and more so that it's easier to understand the codebase. This also helps me find bugs and understand things better :)

### CODE REVIEW

File: lib/AST/NameLookup.cpp [hunk 1]
```diff
@@ -1467,20 +1468,28 @@ DirectLookupRequest::evaluate(Evaluator &evaluator,
...
-    // The lookup table believes it doesn't have a complete accounting of this
-    // name - either because we're never seen it before, or another extension
-    // was registered since the last time we searched. Ask the loaders to give
-    // us a hand.
-    DeclBaseName baseName(name.getBaseName());
-    populateLookupTableEntryFromLazyIDCLoader(ctx, Table, baseName, decl);
-
-    if (!disableAdditionalExtensionLoading) {
-      populateLookupTableEntryFromExtensions(ctx, Table, baseName, decl);
+    if (decl->getClangDecl() &&
```
Comment (zoecarver) [inline lib/AST/NameLookup.cpp:1471]: In the future, I imagine we'll have a switch here, or a lot of if statements, that dispatch out to not just importing namespaces but also records, etc.

File: lib/ClangImporter/ClangImporter.cpp [hunk 2]
```diff
@@ -4005,6 +4004,43 @@ void ClangImporter::Implementation::lookupAllObjCMembers(
...
+      ctx.getClangModuleLoader()->getEffectiveClangContext(declContextTypeDecl);
+
+  return lookupTable->lookup(SerializedSwiftName(desc.name.getBaseName()),
+                             effectiveClangContext);
+}
+
+// REVIEW-QUESTION: does it make sense for this to be its own request? It's
+// modifying global state via the importDeclDirectly call, so that's why I made
+// it a request. If this is the correct approach, I'll pull more of the
+// ImportDecl visitors into their own requests, like I do here for namespaces.
```
Comment (egorzhdan) [inline lib/ClangImporter/ClangImporter.cpp:4028]: This makes perfect sense to me, although I'm not super familiar with the Request API design. If I'm understanding it correctly, the ideal state would be to eventually migrate all of the ClangImporter to Requests, right? I that case, I think it makes sense to have these visitor invocations as separate requests.

**Review by egorzhdan — COMMENTED** [review]
This looks very promising! I think that with this approach the issues when recursive C++ decl importing produces incorrect results will mostly go away.
I've tried importing a few C++ modules from the test suite with this change, and it worked well for me (I left a couple comments where it didn't – feel free to ignore them if you'd like to focus on the bigger picture for now).

It might be too early for this, but I've also tried importing the C++ standard library, and it crashed on `type_traits`, here's the stacktrace:
<details>
<summary>Stacktrace</summary>
  
```
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /Users/egorzh/swift/swift/cmake-build-debug/bin/swift-ide-test --print-module --source-filename x -I . --module-to-print=std.iosfwd --enable-cxx-interop -tools-directory /Users/egorzh/swift/build/Ninja-DebugAssert/llvm-macosx-x86_64/bin --sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
1.	/Users/egorzh/swift/build/Ninja-DebugAssert/llvm-macosx-x86_64/include/c++/v1/type_traits:1364:50: importing 'std::remove_reference'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-ide-test           0x00000001109975ed llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 61
1  swift-ide-test           0x0000000110997b1b PrintStackTraceSignalHandler(void*) + 27
2  swift-ide-test           0x000000011099599b llvm::sys::RunSignalHandlers() + 123
3  swift-ide-test           0x000000011099973d SignalHandler(int) + 205
4  libsystem_platform.dylib 0x00007fff20544d7d _sigtramp + 29
5  libsystem_platform.dylib 0x00007fbbd4aeb4e0 _sigtramp + 18446743784677599104
6  swift-ide-test           0x0000000107f57935 clang::Redeclarable<clang::TagDecl>::getMostRecentDecl() + 21
7  swift-ide-test           0x0000000107f5771e clang::RecordDecl::getMostRecentDecl() + 30
8  swift-ide-test           0x0000000107f57668 clang::CXXRecordDecl::getMostRecentDecl() + 24
9  swift-ide-test           0x000000010eb37275 clang::CXXRecordDecl::getMostRecentNonInjectedDecl() + 21
10 swift-ide-test           0x000000010eb370d8 clang::ClassTemplateSpecializationDecl::getMostRecentDecl() + 24
11 swift-ide-test           0x000000010dacfb68 clang::RedeclarableTemplateDecl::SpecIterator<clang::ClassTemplateSpecializationDecl, clang::RedeclarableTemplateDecl::SpecEntryTraits<clang::ClassTemplateSpecializationDecl>, clang::ClassTemplateSpecializationDecl>::operator*() const + 40
12 swift-ide-test           0x0000000107ff3b45 (anonymous namespace)::SwiftDeclConverter::VisitClassTemplateDecl(clang::ClassTemplateDecl const*) + 133
13 swift-ide-test           0x0000000107fb12e1 clang::declvisitor::Base<llvm::make_const_ptr, (anonymous namespace)::SwiftDeclConverter, swift::Decl*>::Visit(clang::Decl const*) + 817
14 swift-ide-test           0x0000000107fb06bb swift::ClangImporter::Implementation::importDeclImpl(clang::NamedDecl const*, swift::importer::ImportNameVersion, bool&, bool&) + 331
15 swift-ide-test           0x0000000107fb28d2 swift::ClangImporter::Implementation::importDeclAndCacheImpl(clang::NamedDecl const*, swift::importer::ImportNameVersion, bool, bool) + 578
16 swift-ide-test           0x0000000107f28737 swift::ClangImporter::Implementation::importDecl(clang::NamedDecl const*, swift::importer::ImportNameVersion, bool) + 71
17 swift-ide-test           0x0000000107f29333 swift::ClangModuleUnit::getTopLevelDecls(llvm::SmallVectorImpl<swift::Decl*>&) const + 851
18 swift-ide-test           0x0000000107f29d64 swift::ClangModuleUnit::getDisplayDecls(llvm::SmallVectorImpl<swift::Decl*>&) const + 68
19 swift-ide-test           0x00000001052f50e0 swift::ModuleDecl::getDisplayDecls(llvm::SmallVectorImpl<swift::Decl*>&) const + 112
20 swift-ide-test           0x00000001058ec952 swift::ide::printModuleInterface(swift::ModuleDecl*, llvm::ArrayRef<llvm::StringRef>, swift::OptionSet<swift::ide::ModuleTraversal, unsigned int>, swift::ASTPrinter&, swift::PrintOptions const&, bool) + 258
21 swift-ide-test           0x000000010496e215 doPrintModules(swift::CompilerInvocation const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, swift::OptionSet<swift::ide::ModuleTraversal, unsigned int>, swift::PrintOptions const&, bool, bool) + 1301
22 swift-ide-test           0x0000000104966f54 main + 10404
23 libdyld.dylib            0x00007fff2051af5d start + 1
24 libdyld.dylib            0x000000000000000c start + 18446603339973939376
```
</details>

This might be caused by the fact that there's a place in the importer where we still add members to the namespace enums directly, that would probably need to switch to the new `ClangDirectLookupRequest` API that you've added:

https://github.com/apple/swift/blob/d61d9238e706b75bdf02077eaa229e6abea3e104/lib/ClangImporter/ImportDecl.cpp#L4550-L4559


> The Module Printer will now print decls in the __ObjC module as well.

This sounds good to me, although I think it's important to make sure to prevent decls from dependent modules leaking into the generated module interfaces via the __ObjC module (it would probably be nice to have a test for this in the future).

File: lib/IDE/ModuleInterfacePrinting.cpp [hunk 3]
```diff
@@ -654,13 +654,22 @@ void swift::ide::printModuleInterface(
 
       auto *OwningModule = Importer.getClangOwningModule(CN);
       auto I = ClangDecls.find(OwningModule);
-      if (I != ClangDecls.end()) {
+      if (I != ClangDecls.end() &&
+          // Don't re-add this decl if it already exists for "OwningModule".
+          llvm::find_if(I->second, [D](auto p) { return p.first == D; }) ==
+              I->second.end()) {
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/IDE/ModuleInterfacePrinting.cpp:660]: Fair enough. Done.

**Comment by zoecarver** [comment]
Making this a draft again. I ran into some problems that would be solved by making records lazily imported as well, so I'm just going to do that as part of this patch.

File: lib/ClangImporter/ImportDecl.cpp [hunk 4]
```diff
@@ -3524,11 +3421,6 @@ namespace {
       SmallVector<VarDecl *, 4> members;
```
Comment (zoecarver) [inline lib/ClangImporter/ImportDecl.cpp:3421]: In the future we'll be able to refactor and remove _a lot_ of this visitor. But I want to do that in another nfc.

**Comment by zoecarver** [comment]
This still isn't quite ready to be a PR yet. I need to clean up the tests a bit first. But the logic should be ready for review if you want to take another look @egorzhdan.

File: lib/ClangImporter/ImportDecl.cpp (hunk 4 above)
Comment (zoecarver) [reply lib/ClangImporter/ImportDecl.cpp:3425]: I think we will still have to import every member here unless we can make `hasUnreferenceableStorage` lazy, which is unfortunate. But we should be able to remove a lot of the rest of this logic. I'm hoping to move the logic for synthesizing members (such as for unnamed unions) into those member's visitors. That should also help reduce what needs to be done when only importing the class itself.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by MrNeedHelpHere** [comment]
The account u are associating with and others owned were engaged in illegal activities and police reports are being filed do not change anything revert back to how it was..also they are trying to make it to where i can not use or have access to things

**Review by egorzhdan — COMMENTED** [review]
I think this change looks much better now! I haven't seen any crashes so far.

One thing I've noticed is that this code no longer compiles:
```swift
import std

extension std.__1.string {
}
```
The error is:
```
error: 'string' is not a member type of enum '__ObjC.std.__1'
extension std.__1.string {
          ~~~~~~~ ^
__ObjC.std:2:17: note: '__1' declared here
    public enum __1 {
                ^
```

One more issue is that the module interface printer seems to be printing namespace contents from other submodules. This is visible, for example, on `std` submodules & on one of the tests – `test/Interop/Cxx/namespace/Inputs/submodule-a.h` & module `Submodules.SubmoduleA`. The module interface is printed as:
```swift
enum NS1 {
  enum NS2 {
    struct BasicDeepA {
      init()
    }
    struct BasicDeepB {
      init()
    }
  }
  struct BasicB {
    init()
  }
  struct BasicA {
    init()
  }
}
```
Here `BasicB` & `BasicDeepB` shouldn't be printed since they belong to a different submodule (`Submodules.SubmoduleB`).

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: test/api-digester/Outputs/clang-module-dump.txt [hunk 5]
```diff
@@ -568,6 +568,77 @@
           "usr": "c:objc(pl)NSObject"
         }
       ]
+    },
```
Comment (zoecarver) [inline test/api-digester/Outputs/clang-module-dump.txt:571]: Review note: this change is here because we now print the `__ObjC` module.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
Down to only the lldb failures! Progress :)

File: lib/ClangImporter/ImportDecl.cpp [hunk 6]
```diff
@@ -3752,6 +3600,7 @@ namespace {
         }
       }
```
Comment (zoecarver) [inline lib/ClangImporter/ImportDecl.cpp:3602]: Add a TODO comment here like the one you have for the namespace visitor.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: lib/AST/NameLookup.cpp [hunk 7]
```diff
@@ -1448,15 +1450,38 @@ DirectLookupRequest::evaluate(Evaluator &evaluator,
...
-    // name - either because we're never seen it before, or another extension
-    // was registered since the last time we searched. Ask the loaders to give
-    // us a hand.
     DeclBaseName baseName(name.getBaseName());
-    populateLookupTableEntryFromLazyIDCLoader(ctx, Table, baseName, decl);
-    populateLookupTableEntryFromExtensions(ctx, Table, baseName, decl);
 
-    Table.markLazilyComplete(baseName);
+    // If we're building libSyntaxParser, #if out the two clang importer
+    // requests because libSyntaxParser doesn't know about the clang importer.
```
Comment (beccadax) [inline lib/AST/NameLookup.cpp:1456]: Er, what is this comment?

File: lib/AST/ASTPrinter.cpp [hunk 8]
```diff
@@ -2157,10 +2159,51 @@ void PrintAST::printAccessors(const AbstractStorageDecl *ASD) {
   indent();
 }
 
+// This provides logic for looking up all members of a namespace. This is
+// intentionally implemented only in the printer and should *only* be used for
+// debugging.
```
Comment (beccadax) [inline lib/AST/ASTPrinter.cpp:2164]: What do you mean by "should only be used for debugging"? If you ask Xcode for a C++ header's generated interface, it should show how namespace members are imported, shouldn't it?

File: lib/ClangImporter/ImportDecl.cpp [hunk 9]
```diff
@@ -2434,142 +2435,39 @@ namespace {
...
-          auto nd = dyn_cast<clang::NamedDecl>(namespaceDecls.pop_back_val());
-          // Make sure we only import the defenition of a record.
-          if (auto tagDecl = dyn_cast_or_null<clang::TagDecl>(nd))
-            // Some decls, for example ClassTemplateSpecializationDecls, won't
-            // have a definition here. That's OK.
-            nd = tagDecl->getDefinition() ? tagDecl->getDefinition() : tagDecl;
-          if (!nd)
-            continue;
+      ImportedName importedName;
+      std::tie(importedName, std::ignore) = importFullName(decl);
```
Comment (beccadax) [inline lib/ClangImporter/ImportDecl.cpp:2449]: Is there no possibility of the `correctSwiftName` being present or important?

File: lib/AST/NameLookup.cpp [hunk 10]
```diff
@@ -1448,15 +1450,38 @@ DirectLookupRequest::evaluate(Evaluator &evaluator,
...
+    if (decl->getClangDecl() &&
+        isa<clang::NamespaceDecl>(decl->getClangDecl())) {
```
Comment (beccadax) [inline lib/AST/NameLookup.cpp:1458]: Nit: You can use `isa_and_nonnull` for these tests.

**Review by beccadax — APPROVED** [review]
I didn't do a really *deep* review, but everything I noticed seemed pretty minor.

File: lib/AST/NameLookup.cpp (hunk 10 above)
Comment (zoecarver) [reply lib/AST/NameLookup.cpp:1458]: :) thanks, I've needed that several times.

File: lib/AST/ASTPrinter.cpp (hunk 8 above)
Comment (zoecarver) [reply lib/AST/ASTPrinter.cpp:2164]: Fair enough. I'll re-word.

**Comment by zoecarver** [comment]
@beccadax Thank you for the speedy review! I'll apply those comments after the CI passes (or doesn't).

**Comment by zoecarver** [comment]
@swift-ci please test Windows.

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please test Windows

File: lib/ClangImporter/ImportDecl.cpp (hunk 9 above)
Comment (egorzhdan) [reply lib/ClangImporter/ImportDecl.cpp:2449]: I don't think there's a specific reason to ignore it, it's just that compatibility wasn't considered here before.

File: test/Interop/Cxx/operators/member-inline.swift [hunk 11]
```diff
@@ -185,4 +185,13 @@ OperatorsTestSuite.test("PtrToPtr.subscript (inline)") {
   expectEqual(23, arr[0]![0]![0])
 }
 
+// TODO: this causes a crash (does it also crash on main?)
```
Comment (egorzhdan) [inline test/Interop/Cxx/operators/member-inline.swift:188]: Yes, it crashes on main too, it's probably a separate issue.

**Review by egorzhdan — APPROVED** [review]
This looks good!

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
MacOS is passing!!!

@swift-ci please test Windows

**Pull request merged**